scripts/abins/sdata.py

```diff
@@ -135,7 +231,68 @@ def extract(self):
         full_data.update({'frequencies': self._frequencies})
         return full_data
 
-    def check_thresholds(self, return_cases=False, logger=None):
+    def rebin(self, bins: np.array) -> 'SData':
+        """Re-bin the data to a new set of frequency bins
+
+        Data is resampled using np.histogram; no smoothing/interpolation takes
+        place, this is generally intended for moving to a coarser grid.
+
+        Args: New sampling bin edges.
+
+        Returns:
+            A new SData object with resampled data.
+        """
+        old_frequencies = self.get_frequencies()
+        new_frequencies = (bins[:-1] + bins[1:]) / 2
+
+        new_data = {atom_key: {'s':
+                               {order_key: np.histogram(old_frequencies,
+                                                        bins=bins,
+                                                        weights=order_data,
+                                                        density=0)[0]
+                                for order_key, order_data in atom_data['s'].items()}}
+                    for atom_key, atom_data in self._data.items()}
+
+        return self.__class__(data=new_data, frequencies=new_frequencies,
+                              temperature=self.get_temperature(),
+                              sample_form=self.get_sample_form())
+
+    def add_autoconvolution_spectra(self, max_order: Optional[int] = None) -> None:
+        """
+        Atom-by-atom, add higher order spectra by convolution with fundamentals
+
+        Strictly this is only autoconvolution when forming order-2 from order-1;
+        higher orders are formed by repeated convolution with the fundamentals.
+
+        Data should not have been broadened before applying this operation,
+        or this will lead to repeated broadening of higher orders.
+
+        The process will begin with the highest existing order, and repeat until
+        a spectrum of MAX_ORDER is obtained.
+        """
+        if max_order is None:
+            max_order = abins.parameters.autoconvolution['max_order']
+
+        for atom_key, atom_data in self._data.items():
+            for order_index in range(1, max_order + 1):
+                if f'order_{order_index}' in atom_data['s']:
+                    highest_existing_order = order_index
+                else:
+                    break
+            else:
+                # All orders are already full, do nothing
+                return None
```

Comment:
Done, but you should probably take a look at what I did because it has some slightly different behaviours. Hopefully it is also much easier to follow :smile: